api/src/main/java/org/openmrs/validator/PersonValidator.java

```diff
@@ -69,40 +69,80 @@ public void validate(Object target, Errors errors) {
 		
 		Person person = (Person) target;
 		
+		validatePersonNames(person, errors);
+		validatePersonAddresses(person, errors);
+		
+		if (!person.getVoided() && !validatePersonHasAtLeastOneNonVoidedName(person)) {
+			errors.rejectValue("names", "Person.shouldHaveAtleastOneNonVoidedName");
+		}
+		
+		validateBirthDate(errors, person.getBirthdate());
+		validateDeathDate(errors, person.getDeathDate(), person.getBirthdate());
+		if (person.getVoided()) {
+			ValidationUtils.rejectIfEmptyOrWhitespace(errors, "voidReason", "error.null");
+		}
+		
+		validateDeathCause(person, errors);
+		
+		ValidateUtil.validateFieldLengths(errors, Person.class, "gender", "personVoidReason");
+	}
+	
+	/**
+	 * Checks if a person contains any non voided PersonName attribute
+	 *
+	 * @param person The person that is analysed for voided names
+	 * @return true if at leas one personName is not voided
+	 */
+	private boolean validatePersonHasAtLeastOneNonVoidedName(Person person) {
+			return person.getNames().stream().anyMatch(personName -> !personName.getVoided());
+	}
+	
+	/**
+	 * Checks if the names of a person fulfill the expected criteria  
+	 *
+	 * @param person The person whose names should be validated
+	 * @param errors Stores information about errors encountered during validation.
+	 */
+	private void validatePersonNames(Person person, Errors errors) {
 		int index = 0;
-		boolean atLeastOneNonVoidPersonNameLeft = false;
 		for (PersonName personName : person.getNames()) {
 			errors.pushNestedPath("names[" + index + "]");
 			personNameValidator.validate(personName, errors);
-			if (!personName.getVoided()) {
-				atLeastOneNonVoidPersonNameLeft = true;
-			}
 			errors.popNestedPath();
 			index++;
 		}
-		if (!person.getVoided() && !atLeastOneNonVoidPersonNameLeft) {
-			errors.rejectValue("names", "Person.shouldHaveAtleastOneNonVoidedName");
-		}
-		
-		// validate the personAddress
-		index = 0;
+	}
+	
+	/**
+	 * Checks if the addresses of a person fulfill the expected criteria  
+	 *
+	 * @param person The person whose addresses should be validated
+	 * @param errors Stores information about errors encountered during validation.
+	 */
+	private void validatePersonAddresses(Person person, Errors errors) {
+		int index = 0;
 		for (PersonAddress address : person.getAddresses()) {
 			try {
 				errors.pushNestedPath("addresses[" + index + "]");
-				ValidationUtils.invokeValidator(personAddressValidator, address, errors);
+				personAddressValidator.validate(address, errors);
+			}
+			catch (NullPointerException exception) {
```

Comment:
Are you intentionally ignoring the null pointer exception?

Reply:
Yes. Reason (1) To make explicit that a exception is thrown if the value is null. That fact is pretty much hidden by the finally block. I think that overcomplicates the code understanding  (2) To not change the behaviour of the code. I also added a test case to prove:

The issue lies in the fact that personAddressValidator validator does check if the value is null and then throws an exception. That is not how the other classes behave: https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/validator/PersonAddressValidator.java#L67

Reply:
I would avoid changing behavior in a refactoring pull request. Even with the test, you never know what bugs may be introduced which the test does not catch.

Reply:
If I remove the catch() I also have to remove the test case I added. Since it is some weeks ago I forgot that I added this catch() since the newly added test case of me would not work otherwise. Which means I changed the behaviour. I assume you want the test case & try/catch removed?

Reply:
Whatever it takes to restore the original behaviour.